In [32]:
#importing Libraries
import pandas as pd
import numpy as np

In [17]:
data = pd.read_csv("C://Users//vikas//OneDrive//Desktop//Group 9.csv")

In [18]:
data.head()

CustomerID  Age  Income (INR) (in lakhs per annum) Demography  Gender
0       CUS1   25                                  5      Rural    Male
1       CUS2   30                                 10      Urban    Male
2       CUS3   25                                  8      Rural  Female
3       CUS4   40                                 15      Rural  Female
4       CUS5   45                                 20      Urban    Male

In [19]:
# We don't need id.
data.drop(["CustomerID"], axis = 1, inplace = True)
data.head()

Age  Income (INR) (in lakhs per annum) Demography  Gender
0   25                                  5      Rural    Male
1   30                                 10      Urban    Male
2   25                                  8      Rural  Female
3   40                                 15      Rural  Female
4   45                                 20      Urban    Male

In [20]:
#Encoding for Demography
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

data['Demography']=le.fit_transform(data['Demography'])

In [21]:
data.head(2)

Age  Income (INR) (in lakhs per annum)  Demography Gender
0   25                                  5           0   Male
1   30                                 10           1   Male

In [22]:
## Encoding for Gender Using dictionary and map functions

data['Gender']=data['Gender'].map({'Male':0,'Female':1}).astype(int) #Encoding values 
data.head()

Age  Income (INR) (in lakhs per annum)  Demography  Gender
0   25                                  5           0       0
1   30                                 10           1       0
2   25                                  8           0       1
3   40                                 15           0       1
4   45                                 20           1       0

In [23]:
#Feature scaling
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()

In [24]:
data['Age']=sc.fit_transform(data[['Age']])
data['Income (INR) (in lakhs per annum)']=sc.fit_transform(data[['Income (INR) (in lakhs per annum)']])

In [25]:
data

Age  Income (INR) (in lakhs per annum)  Demography  Gender
0 -0.924940                          -1.166880           0       0
1 -0.431638                          -0.286216           1       0
2 -0.924940                          -0.638482           0       1
3  0.554964                           0.594448           0       1
4  1.048265                           1.475113           1       0
5 -1.418241                          -1.166880           1       0
6  1.541566                           1.475113           1       0
7  0.554964                          -0.286216           1       1

In [31]:
#Calculate silhouetee score
from sklearn.metrics import silhouette_score

silhouette_score(data, KMeans(n_clusters =2).fit_predict(data))

C:\Users\vikas\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
C:\Users\vikas\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1440: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


0.42039956815037455

As silhouette score is coming as 0.42039956815037455, which is ranges between 0.25 – 0.5 provides some evidence of the reality of the clusters in the data.

In [11]:
from sklearn.cluster import KMeans

# k=2

In [12]:
kmeans = KMeans(n_clusters =2)
clusters = kmeans.fit_predict(data)

C:\Users\vikas\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)
C:\Users\vikas\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1440: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [13]:
clusters

array([0, 0, 0, 1, 1, 0, 1, 1])

In [14]:
data["Cluster"] = clusters

In [15]:
data

Age  Income (INR) (in lakhs per annum)  Demography  Gender  Cluster
0 -0.924940                          -1.166880           0       0        0
1 -0.431638                          -0.286216           1       0        0
2 -0.924940                          -0.638482           0       1        0
3  0.554964                           0.594448           0       1        1
4  1.048265                           1.475113           1       0        1
5 -1.418241                          -1.166880           1       0        0
6  1.541566                           1.475113           1       0        1
7  0.554964                          -0.286216           1       1        1

As silhouette score is coming as 0.42039956815037455, which is ranges between 0.25 – 0.5 provides some evidence of the reality of the clusters in the data. There are 4 data points in each clusters